### PR TITLE
gateway: Added Get ACL Support, issue: leofs-project/leofs#490

### PR DESCRIPTION
**Pull request by windkit**
## Description
As the first step, reply the ACL of corresponding Bucket

## Related Issue
https://github.com/leo-project/leofs/issues/490

### CODE REVIEW

**Comment by windkit** [comment]
### Use Case
With the PR, `s3cmd` can use remote copy to accelerate the `sync` process when there are identical files

```
$ mkdir test
$ dd if=/dev/urandom of=test/testfile1 bs=4M count=1
1+0 records in
1+0 records out
4194304 bytes (4.2 MB) copied, 0.332763 s, 12.6 MB/s
$ cp test/testfile1 test/testfile2
$ cp test/testfile1 test/testfile3
$ cp test/testfile1 test/testfile4
$ ./s3cmd sync test s3://test/
WARNING: Module python-magic is not available. Guessing MIME types based on file extensions.
upload: 'test/testfile1' -> 's3://test/test/testfile1'  [1 of 1]
 4194304 of 4194304   100% in    0s    34.54 MB/s  done
remote copy: 'test/testfile1' -> 'test/testfile2'
remote copy: 'test/testfile1' -> 'test/testfile3'
remote copy: 'test/testfile1' -> 'test/testfile4'
Done. Uploaded 4194304 bytes in 1.2 seconds, 3.22 MB/s.
```

### Related Issue
https://github.com/leo-project/leofs/issues/639

File: apps/leo_gateway/src/leo_gateway_s3_api.erl [hunk 1]
```diff
@@ -412,6 +412,32 @@ head_bucket(Req, Key, #req_params{access_key_id = AccessKeyId}) ->
...
+                    ?reply_ok(Header, XML, Req);
+                not_found ->
+                    ?reply_not_found([?SERVER_HEADER], Bucket, <<>>, Req);
+                {error, _Cause} ->
+                    ?reply_internal_error([?SERVER_HEADER], Bucket, <<>>, Req)
+            end;
+        {ok, #?METADATA{del = 1}}->
+            ?reply_not_found([?SERVER_HEADER], Key, <<>>, Req);
+        {error, Cause} ->
+            ?reply_fun(Cause, get, Bucket, Key, 0, Req, BeginTime)
```
Comment (mocchira) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: It doesn't make sense to output the access log through reply_fun only when handling internal server error cases.
At first, we have to decide how the access log handled when receiving a GET-ACL request.

@yosukehara 
IMHO, The New Access Log for GET-ACL is needed in addition to existing Access Logs for Bucket/Object operations.
Thoughts?

File: apps/leo_gateway/src/leo_gateway_s3_api.erl [hunk 2]
```diff
@@ -412,6 +412,32 @@ head_bucket(Req, Key, #req_params{access_key_id = AccessKeyId}) ->
...
+
+%% @doc GET-ACL Operation on Objects
+%% As we only support bucket level ACL, reply bucket ACL here
+%% Related Issue: leo-project/leofs#490
+get_object(Req, Key, #req_params{is_acl = true,
+                                 bucket_name = Bucket}) ->
+    BeginTime = leo_date:clock(),
+    case leo_gateway_rpc_handler:head(Key) of
+        {ok, #?METADATA{del = 0}} ->
+            case leo_s3_bucket:find_bucket_by_name(Bucket) of
```
Comment (mocchira) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: leo_s3_bucket:find_bucket_by_name should implement handling obsolete bucket information as 
https://github.com/leo-project/leofs/issues/642 does in order to keep working while both managers are down.
I'll file this as another issue.

File: apps/leo_gateway/src/leo_gateway_s3_api.erl (hunk 1 above)
Comment (yosukehara) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: @mocchira I've checked the current implementation. I agree we add `GET-ACL` in the method column (1st column).

### the format of the current access log:

| Method | Bucket | Key                                                 | # of Children | Size  | Timestamp                         | Unixtime         | Status Code   | Latency | Cache Status |
|--------|--------|-----------------------------------------------------|---------------|-------|-----------------------------------|------------------|---------------|---------|--------------|
|[PUT]   | test   | test/leo_commons/test/leo_mnesia_tests.erl          |  0            |  3444 | 2017-03-23 10:45:15.521332 +0900  | 1490233515521363 | 200           | 1       |   |
|[PUT]   | test   | test/leo_commons/test/leo_net_tests.erl             |  0            |  2483 | 2017-03-23 10:45:15.530438 +0900  | 1490233515530476 | 200           | 1       |   |    
|[GET]   | test   | test/leo_commons/.git/logs/refs/heads/develop       |  0            |  195  | 2017-03-23 10:45:15.533404 +0900  | 1490233515533439 | 200           | 1       | miss |
|[PUT]   | test   | test/leo_commons/.git/logs/refs/remotes/origin/HEAD |  0            |  195  | 2017-03-23 10:45:15.541014 +0900  | 1490233515541036 | 200           | 1       |   |
|[GET]   | test   | test/leo_commons/test/leo_misc_tests.erl            |  0            |  3248 | 2017-03-23 10:46:40.94936 +0900   | 1490233600094965 | 200           | 1       | hit:mem-cache |
|[GET]   | test   | test/leo_commons/test/leo_misc_tests.erl            |  0            |  3248 | 2017-03-23 10:47:21.930926 +0900  | 1490233641930954 | 200           | 1       | hit:mem-cache |

Comment (mocchira) [inline apps/leo_gateway/src/leo_gateway_s3_api.erl]: @yosukehara make sense.
@windkit go forward with this direction.

**Comment by windkit** [comment]
Updated Access Log as

```
[GET-ACL]       test    test/here       0       0       2017-03-23 17:37:25.358631 +0900        1490258245358682        200     2
[GET-ACL]       test    test/not_here   0       0       2017-03-23 17:37:41.783794 +0900        1490258261783861        404     2

```

**Pull request merged**